tests/5.1/error/test_error_severity_warning.c

```diff
@@ -0,0 +1,32 @@
+//===--- test_error_severity_warning.c ----------------------------===//
+//
+//  OpenMP API Version 5.1 Aug 2021
+//
+//  Test uses error directive with message clause to print a message,
+//  if error condition is reached then "runtime error!" is printed,
+//  otherwise is simply used to print a message. 
+//
+////===----------------------------------------------------------------------===//
+
+#include <omp.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include "ompvv.h"
+#include <math.h>
+
+int errors, i;
+
+int test_error_severity_warning() {
+    i = 0;
+    printf("Test should print a 'warning' error message")
+    #pragma omp error severity(warning)
+    i+=5;
+    OMPVV_WARNING_IF("Error directive caused runtime error", i!=5);
+}
```

Comment:
GCC gives the following compile error and warning:
```
test_error_severity_warning.c:21:58: error: expected ‘;’ before ‘#pragma’
   21 |     printf("Test should print a 'warning' error message")
      |                                                          ^
      |                                                          ;
   22 |     #pragma omp error severity(warning)
      |             ~~~                                           
test_error_severity_warning.c:25:1: warning: control reaches end of non-void function [-Wreturn-type]
   25 | }
      | ^
```
Fixing the error, it shows the expected warning:
```
test_error_severity_warning.c:22:13: warning: ‘pragma omp error’ encountered
   22 |     #pragma omp error severity(warning)
      |             ^~~
```

Reply:
@tob2 do you think it would make more sense to just use severity(warning) for each test, instead of changing the makefile? Many tests, not just this one, use multiple directive clauses to ensure one clause works properly, so maybe it makes more sense to just use the warning so that this process doesn't need to be done.

This would cause 2 other issues though,
1. We could still not test for runtime errors, i.e. no way to see if the error warning properly worked
2. Obviously this would not be the case for a severity(fatal) test.

Reply:
I think modifying `run.sh` or the Makefile isn't that bad. There are already some special cases. Besides the environment special cases, there is:
```
# Special rule for test that needs OMPVV lib
$(CURDIR)/tests/4.5/application_kernels/qmcpack_target_static_lib.c.o:
```

If code is added to handle should-fail-at-runtime, based on e.g. a directory name (probably better) or on the file name, it should handle setting the env var and the should-run-time-fail at the same time. (Future proof.)

* * *

I think in the long run, it is not bad to have some support for "should fail". In OpenMP 5.2, quite often the following appears:  "If..., runtime error termination is performed."

In order to test this, having some Makefile support for run-time error terminations would be useful (i.e. testing that it fails with an error).

In OpenMP 5.2, I see the ".., runtime error termination is performed" with:

* If a **map** clause with a **present** _map-type-modifier_ ... list item is not present ...
* .... **device_num.** If the _device-description_ evaluates to **omp_invalid_device**, ...
* If the _device_num,_ _src_device_num,_ or _dst_device_num_ argument of a device memory routine has the value
* **omp_invalid_device,** ...
  *  Likewise for the **omp_pause_resource** routine
* _(And the OpenMP development version has one additional case.)_

And also `OMP_TARGET_OFFLOAD` with mandatory value:
> The **mandatory** value specifies that the effect of any device construct or device memory routine that uses a device that is unavailable or not supported by the implementation, or uses a non-conforming device number, is as if the **omp_invalid_device** device number was used.

which goes alongside with
> [OMP_DEFAULT_DEVICE] ... If _target-offload-var_ is **mandatory** and the number of non-host devices is zero then the
_default-device-var_ is initialized to **omp_invalid_device.** Otherwise ...

I think the last one is not directly testable as it is not known beforehand whether there is an offload device available, but it could be tested in two testcases:
* Some tests, which uses `omp_set_default_device(omp_invalid_device)` – and check for error termination.
* Another test with `OMP_TARGET_OFFLOAD=mandatory` + a check that `if omp_num_devices() > 0)`, omp_get_default_device() is >= 0 and < num_devices – while otherwise, `omp_get_default_device() == omp_invalid_device`.

Reply:
> We could still not test for runtime errors, i.e. no way to see if the error warning properly worked

With a special case in the `Makefile`, there could be a `grep` on the output, searching for the expected output, both for the compile-time and the run-time check, i.e.

*  `2>&1 |grep '...'` and then use the exit status of the grep for fail/pass
* Or, alternatively, to keep the full output:  2>&1 |tee tmp.log;  grep '...' tmp.log && echo PASS || echo FAIL

And just adding an explicit case for the compilation (similar to qmcpack_target_static_lib.c.o) + filtering the testcase from the 'run:' make target — and a similar line for the run-time check. – The only downside if that the explicit grep is needed – but that's only true for the **error** directive.

_For the other cases (... error termination is performed), the run.sh change with a simple should-fail and run.sh changes would be better, as we do not know the content of the error message. – But that's outside of the scope of this pull request._

Alternatively, we could start with some 'special-tag: value' inside the source-code file – and parse those, e.g., by run.sh and honor them (should fail, match a particular string, etc.). That's more flexible, but requires a grep of all source files.

Thoughts?

Reply:
I like the first strategy better, will look into the makefile changes and push a PR. Thank's so much for your helpful advice!